src/solvers/dg_unstructured_quad/dg_2d.jl

```diff
@@ -223,35 +223,65 @@ end
 
 function calc_boundary_flux!(cache, t, boundary_conditions,
                              mesh::UnstructuredQuadMesh, equations, dg::DG)
-  @unpack surface_flux_values = cache.elements
-  @unpack element_id, element_side_id, name  = cache.boundaries
+  @unpack boundary_condition_types, boundary_indices = boundary_conditions
 
-  @threaded for boundary in eachboundary(dg, cache)
-    # Get the element and side IDs on the boundary element
-    element = element_id[boundary]
-    side    = element_side_id[boundary]
+  calc_boundary_flux_by_type!(cache, t, boundary_condition_types, boundary_indices,
+                              mesh, equations, dg)
+  return nothing
+end
 
-    # pull the external state function from the bounary condition dictionary
-    boundary_condition = boundary_conditions[ name[boundary] ]
 
-    # calc boundary flux on the current boundary interface
-    calc_boundary_flux!(surface_flux_values, t, boundary_condition, mesh, equations, dg, cache,
-                        side, element, boundary)
-  end
+# Iterate over tuples of boundary condition types and associated indices
+# in a type-stable way using "lispy tuple programming".
+function calc_boundary_flux_by_type!(cache, t, BCs::NTuple{N,Any},
+                                     BC_indices::NTuple{N,Vector{Int64}},
+                                     mesh::UnstructuredQuadMesh, equations, dg::DG) where {N}
+  # Extract the boundary condition type and index vector
+  boundary_condition = first(BCs)
+  boundary_condition_indices = first(BC_indices)
+  # Extract the remaining types and indices to be processed later
+  remaining_boundary_conditions = Base.tail(BCs)
+  remaining_boundary_condition_indices = Base.tail(BC_indices)
+
+  # process the first boundary condition type
+  calc_boundary_flux!(cache, t, boundary_condition, boundary_condition_indices, mesh, equations, dg)
+
+  # recursively call this method with the unprocessed boundary types
+  calc_boundary_flux_by_type!(cache, t, remaining_boundary_conditions,
+                              remaining_boundary_condition_indices, mesh, equations, dg)
```

Comment:
TBH, I don't really like these fancy lisp-style iterations unless explicitly required for performance reasons. I am not sure if this was benchmarked or not, but it makes the code harder to read and understand (IMHO, almost always does). No objections against using it here, but if we can get away without these constructs except for some extremely critical cases, I'd appreciate it.